wrangler-core/src/main/java/io/cdap/wrangler/utils/ArithmeticOperations.java

```diff
@@ -24,380 +24,483 @@
-      return null;
+    public int getPriority() {
+      return priority;
     }
-    return x.subtract(y).stripTrailingZeros();
   }
 
   /**
-   Arithmetic operation - Multiply a column with another column.
+   * Return the "largest" (e.g. most general) type of the given numbers.
+   * Used for operations applied to columns of multiple types, in order to determine the type of the output column.
+   * The order of increasing generality is Short < Integer < Long < Float < Double < BigDecimal.
+   * Returns null if any input is null.
+   * Throws an error if any input of type Byte is given (this is not currently supported for arithmetic ops).
    */
-  public static Integer multiply(Integer x, Integer y) {
-    if (x == null || y == null) {
-      return null;
-    }
-    return x * y;
-  }
-
-  /**
-   * Arithmetic operation - Multiply a column with another column.
-   */
-  public static Double multiply(Double x, Double y) {
-    if (x == null || y == null) {
-      return null;
-    }
-    return x * y;
-  }
-
-  /**
-   * Arithmetic operation - Multiply a column with another column.
-   */
-  public static Float multiply(Float x, Float y) {
-    if (x == null || y == null) {
-      return null;
-    }
-    return x * y;
-  }
-
-  /**
-   * Arithmetic operation - Multiply a column with another column.
-   */
-  public static BigDecimal multiply(BigDecimal x, BigDecimal y) {
-    if (x == null || y == null) {
-      return null;
-    }
-    return x.multiply(y);
-  }
-
-  /**
-   * Arithmetic operation - Divide a column by another and return the quotient.
-   */
-  public static Integer divideq(Integer x, Integer y) {
-    if (x == null || y == null || y == 0) {
-      return null;
-    }
-    return x / y;
-  }
-
-  /**
-   * Arithmetic operation - Divide a column by another and return the quotient.
-   */
-  public static Double divideq(Double x, Double y) {
-    if (x == null || y == null || y == 0) {
-      return null;
+  private static NumberType getOutputType(Number... nums) throws DirectiveExecutionException {
+    NumberType outputType = NumberType.SHORT;
+    for (Number num : nums) {
+      if (num instanceof Byte) {
+        throw new DirectiveExecutionException("Input cannot be of type 'Byte'.");
+      } else if (num == null) {
+        return null;
+      } else if (num instanceof Integer && NumberType.INTEGER.getPriority() > outputType.getPriority()) {
+        outputType = NumberType.INTEGER;
+      } else if (num instanceof Long && NumberType.LONG.getPriority() > outputType.getPriority()) {
+        outputType = NumberType.LONG;
+      } else if (num instanceof Float && NumberType.FLOAT.getPriority() > outputType.getPriority()) {
+        outputType = NumberType.FLOAT;
+      } else if (num instanceof Double && NumberType.DOUBLE.getPriority() > outputType.getPriority()) {
+        outputType = NumberType.DOUBLE;
+      } else if (num instanceof BigDecimal) {
+        outputType = NumberType.BIGDECIMAL;
```

Comment:
Another potential optimization here is that as soon as the output type is BigDecimal, return. As soon as one of the values is BigDecimal, the output will always be BigDecimal, so no need to keep verifying.

Reply:
We ended up rolling this change back in PR #572, as returning after finding a BigDecimal sometimes resulted in later null values going uncaught, causing a null pointer exception.